pkg/hubble/observer/local_node_watcher.go

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Hubble
+
+package observer
+
+import (
+	"context"
+	"slices"
+
+	flowpb "github.com/cilium/cilium/api/v1/flow"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/stream"
+)
+
+// LocalNodeWatcher populate Hubble flows local node related fields (currently
+// only labels).
+type LocalNodeWatcher struct {
+	mu      lock.Mutex
+	updated int // update counter
+	cache   struct {
+		// labels are represented as a Go map in node.LocalNode, but we need a
+		// key=val slice for Hubble flows.
+		labels []string
+	}
+}
+
+// NewLocalNodeWatcher return a new LocalNodeWatcher. The given context control
+// whether the LocalNodeWatcher gets updated by the localNodeStream. It is safe
+// to use the returned LocalNodeWatcher once the context is cancelled, but its
+// information might be out-of-date.
+func NewLocalNodeWatcher(ctx context.Context, localNodeStream stream.Observable[node.LocalNode]) *LocalNodeWatcher {
+	watcher := LocalNodeWatcher{}
+	localNodeStream.Observe(ctx, watcher.update, watcher.complete)
+	return &watcher
+}
+
+// OnDecodedFlow implements OnDecodedFlow for LocalNodeWatcher. The
+// LocalNodeWatcher populate the flow's node_labels field.
+func (w *LocalNodeWatcher) OnDecodedFlow(_ context.Context, flow *flowpb.Flow) (bool, error) {
+	w.mu.Lock()
+	flow.NodeLabels = w.cache.labels
+	w.mu.Unlock()
+	return false, nil
+}
+
+// update synchronize the LocalNodeWatcher cache with the given LocalNode info.
+func (w *LocalNodeWatcher) update(n node.LocalNode) {
+	labels := sortedLabelSlice(n.Labels)
+	w.mu.Lock()
+	w.cache.labels = labels
+	w.updated++
+	w.mu.Unlock()
+}
+
+// generation return this watcher update counter.
+func (w *LocalNodeWatcher) generation() int {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return w.updated
+}
```

Comment:
does this function get used somewhere?

Reply:
Yes, it is used in the test.

In `NewLocalNodeWatcher`, observing the local node is setup but we get the first emit/update event asynchronously. In other words, the local node watcher cache is probably empty when `NewLocalNodeWatcher` returns. Consequently, if we call `NewLocalNodeWatcher` and check its cache right away we get inconsistent results, so I added a cache update count and the test check it like this:
```go
require.Eventually(t, func() bool {
	return watcher.generation() > 0
}, time.Second, 10*time.Millisecond)
```
The `generation()` func is a helper ensuring to hold the lock.